conans/test/integration/conan_env_test.py

```diff
@@ -9,10 +9,117 @@
 from conans.test.utils.cpp_test_files import cpp_hello_conan_files
 from conans.util.files import load
 from conans import tools
+from nose.plugins.attrib import attr
 
 
 class ConanEnvTest(unittest.TestCase):
 
+    @attr('slow')
+    def shared_in_current_directory_test(self):
+        conanfile = """
+from conans import ConanFile, CMake, tools
+
+
+class LibConan(ConanFile):
+    name = "lib"
+    version = "1.0"
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
+    default_options = "shared=True"
+    generators = "cmake"
+    exports_sources = "*"
+
+    def build(self):
+        cmake = CMake(self)
+        self.run('cmake %s' % cmake.command_line)
+        self.run("cmake --build . %s" % cmake.build_config)
+
+    def package(self):
+        self.copy("*.h", dst="include", src="hello")
+        self.copy("*.lib", dst="lib", keep_path=False)
+        self.copy("*.dll", dst="bin", keep_path=False)
+        self.copy("*.so", dst="lib", keep_path=False)
+        self.copy("*.dylib", dst="lib", keep_path=False)
+        self.copy("*main*", dst="bin", keep_path=False)
+
+"""
+        cmakelists =  """
+project(mytest)
+set(CMAKE_CXX_COMPILER_WORKS 1)
+set(CMAKE_CXX_ABI_COMPILED 1)
+
+SET(CMAKE_SKIP_RPATH 1)
+ADD_LIBRARY(hello SHARED hello.c)
+ADD_EXECUTABLE(main main.c)
+TARGET_LINK_LIBRARIES(main hello)
+"""
+
+        hello_h = """#pragma once
+#ifdef WIN32
+  #define HELLO_EXPORT __declspec(dllexport)
+#else
+  #define HELLO_EXPORT
+#endif
+
+HELLO_EXPORT void hello();
+"""
+
+        client = TestClient()
+        files = {CONANFILE: conanfile,
+                 "CMakeLists.txt": cmakelists,
+                 "hello.c": '#include "hello.h"\nvoid hello(){\nreturn;}',
+                 "hello.h": hello_h,
+                 "main.c": '#include "hello.h"\nint main(){\nhello();\nreturn 0;\n}'}
+
+        client.save(files)
+        client.run("export conan/stable")
+        client.run("install lib/1.0@conan/stable -o lib:shared=True --build missing")
+        client.save({"conanfile.txt": '''
+[requires]
+lib/1.0@conan/stable
+[generators]
+virtualrunenv
+[imports]
+bin, * -> ./bin
+lib, * -> ./bin
+'''}, clean_first=True)
+
+        client.run("install .")
+        # Break possible rpaths built in the exe with absolute paths
+        os.rename(os.path.join(client.current_folder, "bin"),
+                  os.path.join(client.current_folder, "bin2"))
+
+        with tools.chdir(os.path.join(client.current_folder, "bin2")):
+            if platform.system() == "Windows":
+                self.assertEqual(os.system("main.exe"), 0)
+            elif platform.system() == "Darwin":
+                self.assertEqual(os.system("./main"), 0)
+            else:
+                self.assertNotEqual(os.system("./main"), 0)
```

Comment:
So this fails because of not using LD_LIBRARY_PATH?

Reply:
Yes, because Windows and OSX are working fine with the shared in the same directory than the executable, but with Linux, it fails.

Reply:
About your general comment of your review, I don't get it. It's using the standard way to use shared libraries (if there is some standard way to do it):

- There is a library building a shared library
- There is a consumer project importing the shared library (and the executable)
- The consumer tries to execute the imported shared library and executable in the same directory, and it fails in Linux, but works on OSX and WIndows.
- Then I move the shared library to a different directory, and it fails, I'm making sure that there is no harcoded rpaths messing. 
- Finally I use the virtualrunenvironment that declares de LD_LIBRARY_PATH, PATH and DYLD_LIBRARY_PATH to run the executable, and.. magic! it's running agains the shared in the local cache.

I'll put all this comments in the test to help understand it. But I think it's a good test.

Reply:
Ok, perfect now, thanks